### PR TITLE
MGMT-9508: Change copy to singular for SNO Clusters

### PR DESCRIPTION
**Pull request by celdrake**
https://issues.redhat.com/browse/MGMT-9508

Changed the copy of the Assisted Installer to display "Host" instead of "Hosts" for SNO Clusters.

I also unified the functions used to determine if the cluster is an SNO or not. 

See the screen shots to see which texts were changed:

host-copy
host-copy-2

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -38,11 +38,12 @@ yarn install
...
+In order to work with Assisted UI Lib and see changes in Assisted UI, run the 3 commands above
+sequentially. Before using the **yarn start:assisted-ui** script, make sure this project and the
+`assisted-ui` (and optionally the `uhc-portal`) project are located in the same folder.
 
-Next, create a `.env.development.local` file at the root folder of the `assisted-ui` repository and include this variable:
+Next, create a `.env.development.local` file at the root folder of the `assisted-ui` repository and
+include this variable:
 
 ```dotenv
 REACT_APP_API_URL="...ask the team"
```
Comment (batzionb) [inline README.md:49]: since this isn't related to the issue resolved here, better put it in a separate PR

File: src/cim/components/Agent/AgentTable.tsx [hunk 2]
```diff
@@ -16,20 +16,21 @@ import { usePagination } from '../../../common/components/hosts/usePagination';
 
 export const getAgentId = (agent: AgentK8sResource) => agent.metadata?.uid as string;
```
Comment (batzionb) [inline src/cim/components/Agent/AgentTable.tsx:17]: I suggest not to make changes in the CIM code since you can't test it (I'm assuming you don't have ACM set up?) the CIM developers will fix it later

File: src/ocm/components/AddHosts/InventoryAddHost.tsx [hunk 3]
```diff
@@ -28,7 +28,9 @@ const armArchAlert = (
   />
 );
 
-const InventoryAddHosts: React.FC = () => {
+type InventoryAddHostsProps = { isSNO: boolean };
+
+const InventoryAddHosts: React.FC<InventoryAddHostsProps> = ({ isSNO }: InventoryAddHostsProps) => {
```
Comment (batzionb) [inline src/ocm/components/AddHosts/InventoryAddHost.tsx:33]: seem to me it doesn't need to be passed as a property since you have the cluster

File: src/ocm/components/clusterConfiguration/HostInventory.tsx [hunk 4]
```diff
@@ -67,7 +67,7 @@ const HostInventory: React.FC<{ cluster: Cluster }> = ({ cluster }) => {
   );
   const isOpenshiftClusterStorageEnabled = useFeature('ASSISTED_INSTALLER_OCS_FEATURE');
   const isContainerNativeVirtualizationEnabled = useFeature('ASSISTED_INSTALLER_CNV_FEATURE');
-  const isSNO = isSingleNodeCluster(cluster);
+  const isSNOCluster = isSNO(cluster);
```
Comment (batzionb) [inline src/ocm/components/clusterConfiguration/HostInventory.tsx:70]: maybe
const isSno = isSno(cluster)?
less changes and consistent with other places

File: src/ocm/components/clusterConfiguration/discoveryImageModal.tsx [hunk 5]
```diff
@@ -9,30 +9,34 @@ type DiscoveryImageModalButtonProps = {
   ButtonComponent?: typeof Button | typeof ToolbarButton;
   cluster: Cluster;
   idPrefix: string;
+  isSNO: boolean;
+};
+
+type DiscoveryImageModalProps = {
```
Comment (batzionb) [inline src/ocm/components/clusterConfiguration/discoveryImageModal.tsx:15]: since you have the cluster, you can use isSno(cluster) it instead of adding a property

**Comment by batzionb** [comment]
The only overall improvement I can think of is to implement some kind of utility like:
getHostStr(isSno) => isSno ? 'Host' : 'Hosts'
and use it in the string generation, for example
`Add ${getHostStr(isSno)}`
instead of having the condition everywhere

The other things you can see in the comments in the code are basically using isSno(cluster) where the cluster is one of the properties instead of addding an isSno property, and not changing the cim code.

File: src/cim/components/Agent/AgentTable.tsx (hunk 2 above)
Comment (celdrake) [reply src/cim/components/Agent/AgentTable.tsx:17]: Yes, I don't have ACM setup, so I'll revert the changes there. Thanks for pointing it out.

File: src/ocm/components/clusterConfiguration/HostInventory.tsx (hunk 4 above)
Comment (batzionb) [reply src/ocm/components/clusterConfiguration/HostInventory.tsx:70]: you're right, just leave it as is

File: src/common/components/clusterConfiguration/DiscoveryTroubleshootingModal.tsx [hunk 6]
```diff
@@ -13,19 +13,21 @@ import { PrismCode } from '../ui';
 
 export type HostsNotShowingLinkProps = {
   setDiscoveryHintModalOpen: (isOpen: boolean) => void;
+  isSNO?: boolean;
 };
 
 type DiscoveryTroubleshootingModalProps = {
   setDiscoveryHintModalOpen: (isOpen: boolean) => void;
-  isOpen: boolean;
+  isOpen?: boolean;
```
Comment (batzionb) [inline src/common/components/clusterConfiguration/DiscoveryTroubleshootingModal.tsx:21]: Is there a need isOpen changed to optional?

Comment (celdrake) [reply src/common/components/clusterConfiguration/DiscoveryTroubleshootingModal.tsx:21]: Yes, since I reverted the code from CIM which also usies these components, it would result in Linting / TS errors.

Comment (celdrake) [reply src/common/components/clusterConfiguration/DiscoveryTroubleshootingModal.tsx:21]: Sorry, as discussed this was the wrong parameter to mark as optional. I'll fix this :)

**Review by batzionb — DISMISSED** [review]

File: src/common/components/hosts/HostsTable.tsx [hunk 7]
```diff
@@ -13,24 +13,28 @@ import { HostDetail } from '../../../common/components/hosts/HostRowDetail';
...
 };
 
 export const HostsTableEmptyState: React.FC<HostsTableEmptyStateProps> = ({
   setDiscoveryHintModalOpen,
+  isSNO = false,
 }) => (
   <EmptyState
     icon={ConnectedIcon}
-    title="Waiting for hosts..."
+    title={`Waiting for ${getHostStr(isSNO)}...`}
```
Comment (jtomasek) [inline src/common/components/hosts/HostsTable.tsx:31]: Since we already have humanize-plus as a dependency, I am inclined to being explicit and treat the `hosts` string just as another string that needs pluralization and use:
```suggestion
    title={`Waiting for ${pluralize(+isSNO, 'host')}...`}
```

Comment (celdrake) [reply src/common/components/hosts/HostsTable.tsx:31]: @jkilzi also mentioned this library, I will use it instead

File: src/common/components/clusterConfiguration/DiscoveryTroubleshootingModal.tsx [hunk 8]
```diff
@@ -22,10 +24,11 @@ type DiscoveryTroubleshootingModalProps = {
 
 export const HostsNotShowingLink: React.FC<HostsNotShowingLinkProps> = ({
   setDiscoveryHintModalOpen,
+  isSNO = false,
 }) => (
   <Button variant={ButtonVariant.link} onClick={() => setDiscoveryHintModalOpen(true)} isInline>
     <InfoCircleIcon size="sm" />
-    &nbsp;Hosts not showing up?
+    &nbsp;{pluralize(+isSNO, 'Host | Hosts')} not showing up?
```
Comment (jtomasek) [inline src/common/components/clusterConfiguration/DiscoveryTroubleshootingModal.tsx:31]: Did you mean
```suggestion
    &nbsp;{pluralize(+isSNO, 'Host', 'Hosts')} not showing up?
```

The third argument can be omitted as the pluralization defaults to adding 's' at the end.

Comment (celdrake) [reply src/common/components/clusterConfiguration/DiscoveryTroubleshootingModal.tsx]: Yes, I had the wrong syntax. Also I am now using the default as from your suggestion :+1:

File: src/common/components/clusterConfiguration/DiscoveryInstructions.tsx [hunk 9]
```diff
@@ -1,23 +1,31 @@
...
-  showAllInstructions?: boolean;
-}) => (
+  isSNO = false,
+}: DiscoveryInstructionsProps) => (
   <TextContent>
     <Text component="h3">Adding hosts instructions</Text>
     <TextList component="ol" type={OrderType.number} style={{ marginLeft: 0 }}>
-      <TextListItem hidden={!showAllInstructions}>Click the Add hosts button.</TextListItem>
+      <TextListItem hidden={!showAllInstructions}>
+        Click the Add {pluralize(+isSNO, 'host')} button.
```
[the remark this answers is not in the document]
Comment (celdrake) [reply src/common/components/clusterConfiguration/DiscoveryInstructions.tsx:18]: `pluralize` requires the argument to be a number, so the conversion is needed from `isSNO`.

[the remark this answers is not in the document]
Comment (batzionb) [reply src/common/components/clusterConfiguration/DiscoveryInstructions.tsx:18]: @jkilzi pluralize adds an s to the word for any value except for the number 1, including true and false
```
> pluralize(true, 'test')
'tests'
> pluralize(false, 'test')
'tests'
> pluralize(1, 'test')
'test'
> pluralize(0, 'test')
'tests'
```

[the remark this answers is not in the document]
Comment (jkilzi) [reply src/common/components/clusterConfiguration/DiscoveryInstructions.tsx:18]: oh, I understand now. 10x

**Review by jkilzi — APPROVED** [review]
